internal/rpc/otel_interceptor.go

```diff
@@ -0,0 +1,420 @@
+	if err != nil {
+		panic(fmt.Errorf("failed to create responses per rpc metric: %w", err))
+	}
+	return instrumentation{
+		duration:        duration,
+		requestSize:     requestSize,
+		responseSize:    responseSize,
+		requestsPerRPC:  requestsPerRPC,
+		responsesPerRPC: responsesPerRPC,
+	}
+}
+
+func statusCodeAttribute(protocol string, err error) attribute.KeyValue {
+	statusCodeKey := fmt.Sprintf("rpc.%s.status_code", protocol)
+	statusCode := attribute.Int64(statusCodeKey, 0)
+	if err != nil {
+		statusCode = attribute.Int64(statusCodeKey, int64(connect.CodeOf(err)))
+	}
+	return statusCode
+}
+
+// streamingState stores the ongoing metrics for streaming interceptors.
+type streamingState struct {
+	mu              sync.Mutex
+	spec            connect.Spec
+	protocol        string
+	attributes      []attribute.KeyValue
+	error           error
+	sentCounter     int64
+	receivedCounter int64
+	receiveSize     metric.Int64Histogram
+	sendSize        metric.Int64Histogram
+}
+
+// streamingSenderReceiver encapsulates either a StreamingClientConn or a StreamingHandlerConn.
+type streamingSenderReceiver interface {
+	Receive(msg any) error
+	Send(msg any) error
+}
+
+func (s *streamingState) receive(ctx context.Context, msg any, conn streamingSenderReceiver) error {
+	err := conn.Receive(msg)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if errors.Is(err, io.EOF) {
+		return err // nolint:wrapcheck
+	}
+	s.receivedCounter++
+	if err != nil {
+		s.error = err
+		s.attributes = append(s.attributes, statusCodeAttribute(s.protocol, err))
+	}
+	attrs := append(s.attributes, []attribute.KeyValue{ // nolint:gocritic
+		otelMessageTypeAttr.String(otelMessageTypeReceived),
+		otelMessageIDAttr.Int64(s.receivedCounter),
+	}...)
+	if protomsg, ok := msg.(proto.Message); ok {
+		size := proto.Size(protomsg)
+		attrs = append(attrs, otelMessageSizeAttr.Int(size))
+		s.receiveSize.Record(ctx, int64(size), metric.WithAttributes(attrs...))
+	}
+
+	span := trace.SpanFromContext(ctx)
+	span.AddEvent(otelMessageEvent, trace.WithAttributes(attrs...))
+	return err // nolint:wrapcheck
+}
+
+func (s *streamingState) send(ctx context.Context, msg any, conn streamingSenderReceiver) error {
+	err := conn.Send(msg)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if errors.Is(err, io.EOF) {
+		return err // nolint:wrapcheck
+	}
+	s.sentCounter++
+	if err != nil {
+		s.error = err
+		s.attributes = append(s.attributes, statusCodeAttribute(s.protocol, err))
+	}
+	attrs := append(s.attributes, []attribute.KeyValue{ // nolint:gocritic
+		otelMessageTypeAttr.String(otelMessageTypeSent),
+		otelMessageIDAttr.Int64(s.sentCounter),
+	}...)
+	if protomsg, ok := msg.(proto.Message); ok {
+		size := proto.Size(protomsg)
+		attrs = append(attrs, otelMessageSizeAttr.Int(size))
+		s.sendSize.Record(ctx, int64(size), metric.WithAttributes(attrs...))
+	}
+
+	span := trace.SpanFromContext(ctx)
+	span.AddEvent(otelMessageEvent, trace.WithAttributes(attrs...))
+	return err // nolint:wrapcheck
```

Comment:
no we want to `nolint` all these `err` or should we wrap them?

Reply:
Good question; the interceptor just logs errors as they pass through, and it didn't feel quite right to attach this context to the trace; all errors that flow through rpc would then have some `otel error:` string attached